L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorSuperTriggerCellImpl.cc

```diff
@@ -86,53 +123,310 @@ HGCalConcentratorSuperTriggerCellImpl::getSuperTriggerCellId(int detid) const {
+            
+            trigCellVecOutput.push_back ( triggerCell );
+            
+          }
+          
+        }
+        
+      }
+      
+
+      else if (stcSize_.at(thickness) == 8){
+        
+        for ( int i = 0; i < 2; i++ ){
+          for ( int j = 0; j < 2; j++ ){
+            for ( int k = 0; k < 2; k++ ){
+              
+              satisfy = false;
+              for (auto tc:s.second.GetTCList() ){
+                HGCalDetId TC_idV8(tc);
+                if ( ((TC_idV8.cell() & 1) == i && ((TC_idV8.cell()>>1) & 1) == j && ((TC_idV8.cell()>>2) & 1) == k)){
+                  satisfy = true;
+                }
+              }
+              if ( satisfy == false ){//Create new TC
+                int tc_base = s.second.GetTCList().at(0);
+                //Clear relevant bits       
+                tc_base = tc_base & ~(1); 
+                tc_base = tc_base & ~(1 << 1); 
+                tc_base = tc_base & ~(1 << 2); 
+                //Set bits based on i value
+                int newtc = tc_base | i;
+                newtc = newtc | (j<<1);
+                newtc = newtc | (k<<2);
+                
+                if ( !triggerTools_.validTriggerCell(newtc) ) {
+                  s.second.reject();
+                  continue;
+                }
+                
+                
+                l1t::HGCalTriggerCell triggerCell;
+                GlobalPoint point = triggerTools_.getTCPosition(newtc);
+                math::PtEtaPhiMLorentzVector p4(0, point.eta(), point.phi(), 0.);
+                
+                triggerCell.setPosition(point);     
+                triggerCell.setP4(p4);
+                triggerCell.setDetId(newtc);
+
+                s.second.addToList( triggerCell );                      
+                trigCellVecOutput.push_back ( triggerCell );
+                
+              }
+              
+            }
+            
+          }
+          
+        }
+        
+      }
+     
+
+ 
+    }
+
+        
+ 
+}
+
+
+void 
+HGCalConcentratorSuperTriggerCellImpl::
+coarsenTriggerCells( std::unordered_map<unsigned,SuperTriggerCell>& STCs, const std::vector<l1t::HGCalTriggerCell>& trigCellVecInput, std::vector<l1t::HGCalTriggerCell>& trigCellVecOutput)
+{ 
+  std::unordered_map<unsigned,SuperTriggerCell> coarseTCs;
+
+
+  for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
+    if (tc.subdetId() != HGCHEB) {
+      coarseTCs[getSuperTriggerCellId(tc.detId(),2)].add(tc);
+    }
+    else{
+      trigCellVecOutput.push_back( tc );
+    }
+  }
+
+  for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
+     if (tc.subdetId() == HGCHEB) continue;
+
+     int thickness = triggerTools_.thicknessIndex(tc.detId(),true);
+     if ( thickness > 0 ){
+
+      if (!(tc.detId() & 1)) { //if even
+        const auto & ctc = coarseTCs[getSuperTriggerCellId(tc.detId(),2)]; 
+        auto & stc = STCs[getSuperTriggerCellId(tc.detId())]; 
```

Comment:
It's not clear why both a "coarse TC" and a "super TC" need to be created. To my understanding, a super TC is a particular type of coarse TC, so I would expect a single object type that could handle all the particular cases.

Reply:
Right, but in the fixed-data size proposal, we have super trigger cells being formed of coarse TCs (in the thick layers). Therefore we need a separate list.

Reply:
But in that case "coarse TCs" would be normal TC objects? Where would they be constructed, in the HGCROC or in the ECON? Sorry I probably missed some parts of the Coarse TC bestiary. It will probably be more efficient to discuss directly on Skype.

Reply:
The coarse TC solution is somewhat temporary, before we redefine the which sensor cells go into the TCs in the thicker layers. So probably we do need a separate coarseTC and superTC here at the moment